File: RecyclerViews/RecyclerViewDemo3/app/src/main/java/edu/cs4730/recyclerviewdemo3/Phonebook_myAdapter.java
package edu.cs4730.recyclerviewdemo3;

import java.util.List;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import edu.cs4730.recyclerviewdemo3.databinding.PhonebookRowlayoutBinding;

/**
 *  needs a comment here.
 */

public class Phonebook_myAdapter extends RecyclerView.Adapter<Phonebook_myAdapter.ViewHolder> implements OnClickListener {
    private Context mContext;
    private int rowLayout;
    private List<Phonebook_DataModel> listPhonebook;

    public Phonebook_myAdapter(List<Phonebook_DataModel> listPhonebook, int rowLayout, Context context) {
        this.mContext = context;
        this.listPhonebook = listPhonebook;
        this.rowLayout = rowLayout;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        PhonebookRowlayoutBinding v = PhonebookRowlayoutBinding.inflate(LayoutInflater.from(mContext), viewGroup, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int position) {
        //find our place in the datamodel and recycler view.
        Phonebook_DataModel entry = listPhonebook.get(position);
        //setup the view with data.
        viewHolder.viewBinding.tvContact.setText(entry.getName());
        viewHolder.viewBinding.tvMobile.setText(entry.getPhone());
        viewHolder.viewBinding.tvMail.setText(entry.getMail());

        // Set the onClick Listener on this button
        viewHolder.viewBinding.btnRemove.setOnClickListener(this);
        viewHolder.viewBinding.btnRemove.setFocusableInTouchMode(false);
        viewHolder.viewBinding.btnRemove.setFocusable(false);
        // Set the entry, so that you can capture which item was clicked and
        // then remove it
        // As an alternative, you can use the id/position of the item to capture
        // the item that was clicked.
        // btnRemove.setId(position);
        viewHolder.viewBinding.btnRemove.setTag(entry);

    }

    @Override
    public int getItemCount() {
        return listPhonebook == null ? 0 : listPhonebook.size();
    }

    /*
     * setup the ViewHolder class with the widget variables, to be used in onBindViewholder
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        PhonebookRowlayoutBinding viewBinding;
        public ViewHolder(PhonebookRowlayoutBinding itemView) {
            super(itemView.getRoot());
            viewBinding = itemView;
        }
    }


    @Override
    public void onClick(View view) {
        Phonebook_DataModel entry = (Phonebook_DataModel) view.getTag();
        //We could call a dialog showDialog(entry), if we wanted to change it instead of deleting it.

        listPhonebook.remove(entry);
        notifyDataSetChanged();

    }

    private void showDialog(Phonebook_DataModel entry) {
        // Create and show your dialog
        // Depending on the Dialogs button clicks delete it or do nothing
    }


}

File: RecyclerViews/RecyclerViewDemo3_kt/app/src/main/java/edu/cs4730/recyclerviewdemo3_kt/Phonebook_myAdapter.kt
package edu.cs4730.recyclerviewdemo3_kt

import android.content.Context
import edu.cs4730.recyclerviewdemo3_kt.Phonebook_DataModel
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.TextView
import edu.cs4730.recyclerviewdemo3_kt.R
import edu.cs4730.recyclerviewdemo3_kt.databinding.PhonebookRowlayoutBinding

/**
 * needs a comment here.
 */
class Phonebook_myAdapter(
    private val listPhonebook: MutableList<Phonebook_DataModel>?,
    private val mContext: Context
) : RecyclerView.Adapter<Phonebook_myAdapter.ViewHolder>(), View.OnClickListener {
    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val v = PhonebookRowlayoutBinding.inflate(LayoutInflater.from(mContext), viewGroup, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        //find our place in the datamodel and recycler view.
        val entry = listPhonebook!![position]
        //setup the view with data.
        viewHolder.viewBinding.tvContact.text = entry.name
        viewHolder.viewBinding.tvMobile.text = entry.phone
        viewHolder.viewBinding.tvMail.text = entry.mail

        // Set the onClick Listener on this button
        viewHolder.viewBinding.btnRemove.setOnClickListener(this)
        viewHolder.viewBinding.btnRemove.isFocusableInTouchMode = false
        viewHolder.viewBinding.btnRemove.isFocusable = false
        // Set the entry, so that you can capture which item was clicked and
        // then remove it
        // As an alternative, you can use the id/position of the item to capture
        // the item that was clicked.
        // btnRemove.setId(position);

        viewHolder.viewBinding.btnRemove.tag = entry
    }

    override fun getItemCount(): Int {
        return listPhonebook?.size ?: 0
    }

    //viewbinding provides the references
    inner class ViewHolder(var viewBinding: PhonebookRowlayoutBinding) :
        RecyclerView.ViewHolder(viewBinding.root) {
    }

    override fun onClick(view: View) {
        val entry = view.tag as Phonebook_DataModel
        //We could call a dialog showDialog(entry), if we wanted to change it instead of deleting it.
        listPhonebook!!.remove(entry)
        notifyDataSetChanged()
    }

    private fun showDialog(entry: Phonebook_DataModel) {
        // Create and show your dialog
        // Depending on the Dialogs button clicks delete it or do nothing
    }
}